<!DOCTYPE html>
<html lang="zh-cn">
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <title>Python教程065 异常 - None - 记录在日本从事IT的所感所想。</title>
  <meta name="renderer" content="webkit" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>

<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />

<meta name="theme-color" content="#f8f5ec" />
<meta name="msapplication-navbutton-color" content="#f8f5ec">
<meta name="apple-mobile-web-app-capable" content="yes">
<meta name="apple-mobile-web-app-status-bar-style" content="#f8f5ec">


<meta name="author" content="inuwashi" /><meta name="description" content="异常 一般情况下，在Python无法正常处理程序时就会发生一个异常。异常是Python对象，在python中，所有异常都必须是继承BaseEx" /><meta name="keywords" content="日本IT, 程序开发, 运维保守" />






<meta name="generator" content="Hugo 0.57.2 with even 4.0.0" />


<link rel="canonical" href="https://inuwashi123.github.io/post/python/python%E6%95%99%E7%A8%8B065-%E5%BC%82%E5%B8%B8/" />
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="manifest" href="/manifest.json">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">


<link href="/dist/even.c2a46f00.min.css" rel="stylesheet">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.1.20/dist/jquery.fancybox.min.css" integrity="sha256-7TyXnr2YU040zfSP+rEcz29ggW4j56/ujTPwjMzyqFY=" crossorigin="anonymous">


<meta property="og:title" content="Python教程065 异常" />
<meta property="og:description" content="异常 一般情况下，在Python无法正常处理程序时就会发生一个异常。异常是Python对象，在python中，所有异常都必须是继承BaseEx" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://inuwashi123.github.io/post/python/python%E6%95%99%E7%A8%8B065-%E5%BC%82%E5%B8%B8/" />
<meta property="article:published_time" content="2019-11-21T09:09:44+09:00" />
<meta property="article:modified_time" content="2019-11-21T09:09:44+09:00" />
<meta itemprop="name" content="Python教程065 异常">
<meta itemprop="description" content="异常 一般情况下，在Python无法正常处理程序时就会发生一个异常。异常是Python对象，在python中，所有异常都必须是继承BaseEx">


<meta itemprop="datePublished" content="2019-11-21T09:09:44&#43;09:00" />
<meta itemprop="dateModified" content="2019-11-21T09:09:44&#43;09:00" />
<meta itemprop="wordCount" content="2200">



<meta itemprop="keywords" content="python教程," />
<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Python教程065 异常"/>
<meta name="twitter:description" content="异常 一般情况下，在Python无法正常处理程序时就会发生一个异常。异常是Python对象，在python中，所有异常都必须是继承BaseEx"/>

<!--[if lte IE 9]>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/classlist/1.1.20170427/classList.min.js"></script>
<![endif]-->

<!--[if lt IE 9]>
  <script src="https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js"></script>
<![endif]-->

</head>
<body>
  <div id="mobile-navbar" class="mobile-navbar">
  <div class="mobile-header-logo">
    <a href="/" class="logo">None</a>
  </div>
  <div class="mobile-navbar-icon">
    <span></span>
    <span></span>
    <span></span>
  </div>
</div>
<nav id="mobile-menu" class="mobile-menu slideout-menu">
  <ul class="mobile-menu-list">
    <a href="/">
        <li class="mobile-menu-item">Home</li>
      </a><a href="/post/">
        <li class="mobile-menu-item">Archives</li>
      </a><a href="/tags/">
        <li class="mobile-menu-item">Tags</li>
      </a><a href="/categories/">
        <li class="mobile-menu-item">Categories</li>
      </a><a href="/about/">
        <li class="mobile-menu-item">About</li>
      </a>
  </ul>
</nav>
  <div class="container" id="mobile-panel">
    <header id="header" class="header">
        <div class="logo-wrapper">
  <a href="/" class="logo">None</a>
</div>

<nav class="site-navbar">
  <ul id="menu" class="menu">
    <li class="menu-item">
        <a class="menu-item-link" href="/">Home</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/post/">Archives</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/tags/">Tags</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/categories/">Categories</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/about/">About</a>
      </li>
  </ul>
</nav>
    </header>

    <main id="main" class="main">
      <div class="content-wrapper">
        <div id="content" class="content">
          <article class="post">
    
    <header class="post-header">
      <h1 class="post-title">Python教程065 异常</h1>

      <div class="post-meta">
        <span class="post-time"> 2019-11-21 </span>
        <div class="post-category">
            <a href="/categories/python/"> python </a>
            </div>
          <span class="more-meta"> 约 2200 字 </span>
          <span class="more-meta"> 预计阅读 5 分钟 </span>
        
      </div>
    </header>

    <div class="post-toc" id="post-toc">
  <h2 class="post-toc-title">文章目录</h2>
  <div class="post-toc-content">
    <nav id="TableOfContents">
<ul>
<li>
<ul>
<li><a href="#异常">异常</a>
<ul>
<li><a href="#异常处理">异常处理</a></li>
<li><a href="#抛出异常">抛出异常</a></li>
<li><a href="#处理多个异常">处理多个异常</a></li>
<li><a href="#异常参数">异常参数</a></li>
</ul></li>
</ul></li>
</ul>
</nav>
  </div>
</div>
    <div class="post-content">
      

<h2 id="异常">异常</h2>

<p>一般情况下，在Python无法正常处理程序时就会发生一个异常。异常是Python对象，在python中，所有异常都必须是继承BaseException类的实例。当Python脚本发生异常时我们需要捕获处理它，否则程序会终止执行。</p>

<h3 id="异常处理">异常处理</h3>

<p>捕捉异常可以使用try&hellip;except&hellip;else&hellip;finally&hellip;语句。其中else和finally子句是可以省略的。格式如下，</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span><span class="lnt">6
</span><span class="lnt">7
</span><span class="lnt">8
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="k">try</span><span class="p">:</span>
    <span class="err">语句块</span>
<span class="k">except</span> <span class="err">异常表达式</span><span class="p">:</span>
    <span class="n">except处理语句块</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">else语句块</span>
<span class="k">finally</span><span class="p">:</span>
    <span class="n">finally语句块</span></code></pre></td></tr></table>
</div>
</div>
<p>按照如下方式工作；<br />
首先，执行try子句（在关键字try和关键字except之间的语句）
如果没有异常发生，忽略except子句，try子句执行完后，依次执行else子句和finally子句后结束。</p>

<p>如果在执行try子句的过程中发生了异常，那么try子句余下的部分将被忽略。如果异常的类型和 except 之后的名称相符，那么对应的except子句将被执行。最后执行finally子句。</p>

<p>最后一个except子句,可以没有异常表达式，表示它匹配任何异常。
如果一个异常没有与任何的except匹配，那么这个异常将会传递给上层的try中。</p>

<p>一个 try语句可能包含多个except子句，分别来处理不同的特定的异常。最多只有一个分支会被执行。</p>

<p>else子句必须放在所有的 except 子句后面。else从句只会在没有异常的情况下执行，而且它会在finally语句之前执行。</p>

<p>finally字句是在最后一定会被执行的。</p>

<p>例如：</p>

<p>try&hellip;except&hellip;语句，捕获不能被零除的异常。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span><span class="lnt">6
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="k">try</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;try...&#39;</span><span class="p">)</span>
    <span class="n">r</span> <span class="o">=</span> <span class="mi">10</span> <span class="o">/</span> <span class="mi">0</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;result:&#39;</span><span class="p">,</span> <span class="n">r</span><span class="p">)</span>
<span class="k">except</span> <span class="ne">ZeroDivisionError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;except:&#39;</span><span class="p">,</span> <span class="n">e</span><span class="p">)</span></code></pre></td></tr></table>
</div>
</div>
<p>执行以上程序会输出如下结果：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span><span class="lnt">2
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="k">try</span><span class="o">...</span>
<span class="k">except</span><span class="p">:</span> <span class="n">division</span> <span class="n">by</span> <span class="n">zero</span></code></pre></td></tr></table>
</div>
</div>
<p>try&hellip;except&hellip;finally&hellip;语句，捕获不能被零除的异常。无论是否有异常，finally子句总是最后被执行。</p>

<p>例如：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span><span class="lnt">6
</span><span class="lnt">7
</span><span class="lnt">8
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="k">try</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;try...&#39;</span><span class="p">)</span>
    <span class="n">r</span> <span class="o">=</span> <span class="mi">10</span> <span class="o">/</span> <span class="mi">0</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;result:&#39;</span><span class="p">,</span> <span class="n">r</span><span class="p">)</span>
<span class="k">except</span> <span class="ne">ZeroDivisionError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;except:&#39;</span><span class="p">,</span> <span class="n">e</span><span class="p">)</span>
<span class="k">finally</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;finally&#39;</span><span class="p">)</span></code></pre></td></tr></table>
</div>
</div>
<p>执行以上程序会输出如下结果：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="k">try</span><span class="o">...</span>
<span class="k">except</span><span class="p">:</span> <span class="n">division</span> <span class="n">by</span> <span class="n">zero</span>
<span class="k">finally</span></code></pre></td></tr></table>
</div>
</div>
<p>try&hellip;except&hellip;else&hellip;finally&hellip;语句语句，捕获不能被零除的异常。有异常时，else子句不会执行。但是finally子句总是最后被执行。</p>

<p>例如：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="k">try</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;try...&#39;</span><span class="p">)</span>
    <span class="n">r</span> <span class="o">=</span> <span class="mi">10</span> <span class="o">/</span> <span class="mi">0</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;result:&#39;</span><span class="p">,</span> <span class="n">r</span><span class="p">)</span>
<span class="k">except</span> <span class="ne">ZeroDivisionError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;except:&#39;</span><span class="p">,</span> <span class="n">e</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;else&#39;</span><span class="p">)</span>
<span class="k">finally</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;finally&#39;</span><span class="p">)</span></code></pre></td></tr></table>
</div>
</div>
<p>执行以上程序会输出如下结果：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="k">try</span><span class="o">...</span>
<span class="k">except</span><span class="p">:</span> <span class="n">division</span> <span class="n">by</span> <span class="n">zero</span>
<span class="k">finally</span></code></pre></td></tr></table>
</div>
</div>
<p>try&hellip;except&hellip;else&hellip;finally&hellip;语句语句，捕获不能被零除的异常。无异常时，else子句会执行。finally子句总是最后被执行。</p>

<p>例如：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="k">try</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;try...&#39;</span><span class="p">)</span>
    <span class="n">r</span> <span class="o">=</span> <span class="mi">10</span> <span class="o">/</span> <span class="mi">1</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;result:&#39;</span><span class="p">,</span> <span class="n">r</span><span class="p">)</span>
<span class="k">except</span> <span class="ne">ZeroDivisionError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;except:&#39;</span><span class="p">,</span> <span class="n">e</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;else&#39;</span><span class="p">)</span>
<span class="k">finally</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;finally&#39;</span><span class="p">)</span></code></pre></td></tr></table>
</div>
</div>
<p>执行以上程序会输出如下结果：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="k">try</span><span class="o">...</span>
<span class="n">result</span><span class="p">:</span> <span class="mf">10.0</span>
<span class="k">else</span>
<span class="k">finally</span></code></pre></td></tr></table>
</div>
</div>
<h3 id="抛出异常">抛出异常</h3>

<p>Python使用raise语句抛出一个指定的异常。</p>

<p>例如：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="k">try</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">NameError</span><span class="p">(</span><span class="s1">&#39;NoExistValue&#39;</span><span class="p">)</span>
<span class="k">except</span> <span class="ne">NameError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;except:&#39;</span><span class="p">,</span> <span class="n">e</span><span class="p">)</span></code></pre></td></tr></table>
</div>
</div>
<p>执行以上程序会输出如下结果：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="k">except</span><span class="p">:</span> <span class="n">NoExistValue</span></code></pre></td></tr></table>
</div>
</div>
<h3 id="处理多个异常">处理多个异常</h3>

<p>可以使用三种方法来处理多个异常。
第一种方法需要把所有可能发生的异常放到一个元组。</p>

<p>例如：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="k">try</span><span class="p">:</span>
    <span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;test.txt&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<span class="k">except</span> <span class="p">(</span><span class="ne">IOError</span><span class="p">,</span> <span class="ne">EOFError</span><span class="p">)</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s2">&#34;An error occurred. {}&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">args</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span></code></pre></td></tr></table>
</div>
</div>
<p>执行以上程序会输出如下结果：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="n">An</span> <span class="n">error</span> <span class="n">occurred</span><span class="o">.</span> <span class="p">[</span><span class="n">Errno</span> <span class="mi">2</span><span class="p">]</span> <span class="n">No</span> <span class="n">such</span> <span class="nb">file</span> <span class="ow">or</span> <span class="n">directory</span><span class="p">:</span> <span class="s1">&#39;test.txt&#39;</span></code></pre></td></tr></table>
</div>
</div>
<p>第二种方式是对每个单独的异常在单独的except语句块中处理。我们想要多少个except语句块都可以。</p>

<p>例如：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span><span class="lnt">6
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="k">try</span><span class="p">:</span>
    <span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;test.txt&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<span class="k">except</span> <span class="ne">EOFError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s2">&#34;An EOF error occurred.&#34;</span><span class="p">,</span><span class="n">e</span><span class="p">)</span>
<span class="k">except</span> <span class="ne">IOError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s2">&#34;An error occurred.&#34;</span><span class="p">,</span><span class="n">e</span><span class="p">)</span></code></pre></td></tr></table>
</div>
</div>
<p>执行以上程序会输出如下结果：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="n">An</span> <span class="n">error</span> <span class="n">occurred</span><span class="o">.</span> <span class="p">[</span><span class="n">Errno</span> <span class="mi">2</span><span class="p">]</span> <span class="n">No</span> <span class="n">such</span> <span class="nb">file</span> <span class="ow">or</span> <span class="n">directory</span><span class="p">:</span> <span class="s1">&#39;test.txt&#39;</span></code></pre></td></tr></table>
</div>
</div>
<p>第三种方式是except 子句可以省略异常名，以用作通配符。捕获所有异常。</p>

<p>例如：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span><span class="lnt">6
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="k">try</span><span class="p">:</span>
    <span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;test.txt&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<span class="k">except</span> <span class="ne">EOFError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s2">&#34;An EOF error occurred.&#34;</span><span class="p">,</span><span class="n">e</span><span class="p">)</span>
<span class="k">except</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s2">&#34;An error occurred.&#34;</span><span class="p">)</span></code></pre></td></tr></table>
</div>
</div>
<p>执行以上程序会输出如下结果：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="n">An</span> <span class="n">error</span> <span class="n">occurred</span><span class="o">.</span></code></pre></td></tr></table>
</div>
</div>
<p>如果发生的异常和except子句中的类是同一个类或者是except子句中的类的子类，则异常和except子句中的类是兼容的（但反过来则不成立）。</p>

<p>例如，下面的代码将依次打印 Grandpa, Father, Son</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="k">class</span> <span class="nc">Grandpa</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="k">pass</span>

<span class="k">class</span> <span class="nc">Father</span><span class="p">(</span><span class="n">Grandpa</span><span class="p">):</span>
    <span class="k">pass</span>

<span class="k">class</span> <span class="nc">Son</span><span class="p">(</span><span class="n">Father</span><span class="p">):</span>
    <span class="k">pass</span>

<span class="k">for</span> <span class="n">instance</span> <span class="ow">in</span> <span class="p">[</span><span class="n">Grandpa</span><span class="p">,</span> <span class="n">Father</span><span class="p">,</span> <span class="n">Son</span><span class="p">]:</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">instance</span><span class="p">()</span>
    <span class="k">except</span> <span class="n">Son</span><span class="p">:</span>
        <span class="k">print</span><span class="p">(</span><span class="s2">&#34;Son&#34;</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">Father</span><span class="p">:</span>
        <span class="k">print</span><span class="p">(</span><span class="s2">&#34;Father&#34;</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">Grandpa</span><span class="p">:</span>
        <span class="k">print</span><span class="p">(</span><span class="s2">&#34;Grandpa&#34;</span><span class="p">)</span></code></pre></td></tr></table>
</div>
</div>
<p>执行以上程序会输出如下结果：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="n">Grandpa</span>
<span class="n">Father</span>
<span class="n">Son</span></code></pre></td></tr></table>
</div>
</div>
<p>但是，下面的代码依次打印Grandpa, Grandpa, Grandpa</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="k">class</span> <span class="nc">Grandpa</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="k">pass</span>

<span class="k">class</span> <span class="nc">Father</span><span class="p">(</span><span class="n">Grandpa</span><span class="p">):</span>
    <span class="k">pass</span>

<span class="k">class</span> <span class="nc">Son</span><span class="p">(</span><span class="n">Father</span><span class="p">):</span>
    <span class="k">pass</span>

<span class="k">for</span> <span class="n">instance</span> <span class="ow">in</span> <span class="p">[</span><span class="n">Grandpa</span><span class="p">,</span> <span class="n">Father</span><span class="p">,</span> <span class="n">Son</span><span class="p">]:</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">instance</span><span class="p">()</span>
    <span class="k">except</span> <span class="n">Grandpa</span><span class="p">:</span>
        <span class="k">print</span><span class="p">(</span><span class="s2">&#34;Grandpa&#34;</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">Son</span><span class="p">:</span>
        <span class="k">print</span><span class="p">(</span><span class="s2">&#34;Son&#34;</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">Father</span><span class="p">:</span>
        <span class="k">print</span><span class="p">(</span><span class="s2">&#34;Father&#34;</span><span class="p">)</span></code></pre></td></tr></table>
</div>
</div>
<p>执行以上程序会输出如下结果：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="n">Grandpa</span>
<span class="n">Grandpa</span>
<span class="n">Grandpa</span></code></pre></td></tr></table>
</div>
</div>
<h3 id="异常参数">异常参数</h3>

<p>发生异常时，它可能具有关联值，显示异常的相关信息，也称为异常参数 。参数的存在和类型取决于异常类型。</p>

<p>except 子句可以在异常名称后面指定一个变量。这个变量和一个异常实例绑定，它的参数存储在 instance.args 中。为了方便起见，异常实例定义了<strong>str</strong>() ，因此可以直接打印参数而无需引用 .args 。也可以在抛出之前首先实例化异常，并根据需要向其添加任何属性。</p>

<p>例如：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="k">try</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s1">&#39;spam&#39;</span><span class="p">,</span> <span class="s1">&#39;eggs&#39;</span><span class="p">)</span>
<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">inst</span><span class="p">:</span>
    <span class="c1"># 异常实例的类型</span>
    <span class="k">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">inst</span><span class="p">))</span>

    <span class="c1"># 异常参数arguments stored in .args</span>
    <span class="k">print</span><span class="p">(</span><span class="n">inst</span><span class="o">.</span><span class="n">args</span><span class="p">)</span>

    <span class="c1"># __str__方法允许直接打印inst</span>
    <span class="k">print</span><span class="p">(</span><span class="n">inst</span><span class="p">)</span>

    <span class="c1"># 异常参数解包</span>
    <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">inst</span><span class="o">.</span><span class="n">args</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;x =&#39;</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;y =&#39;</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span></code></pre></td></tr></table>
</div>
</div>
<p>执行以上程序会输出如下结果：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="o">&lt;</span><span class="k">class</span> <span class="err">&#39;</span><span class="nc">Exception</span><span class="s1">&#39;&gt;</span>
<span class="p">(</span><span class="s1">&#39;spam&#39;</span><span class="p">,</span> <span class="s1">&#39;eggs&#39;</span><span class="p">)</span>
<span class="p">(</span><span class="s1">&#39;spam&#39;</span><span class="p">,</span> <span class="s1">&#39;eggs&#39;</span><span class="p">)</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">spam</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">eggs</span></code></pre></td></tr></table>
</div>
</div>
<p>异常处理程序不仅处理 try 子句中遇到的异常，还处理 try 子句中调用（即使是间接地）的函数内部发生的异常。</p>

<p>例如：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span><span class="lnt">6
</span><span class="lnt">7
</span><span class="lnt">8
</span><span class="lnt">9
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="c1">#定义引发异常的函数</span>
<span class="k">def</span> <span class="nf">this_fails</span><span class="p">():</span>
    <span class="n">x</span> <span class="o">=</span> <span class="mi">1</span><span class="o">/</span><span class="mi">0</span>

<span class="k">try</span><span class="p">:</span>
    <span class="c1">#调用函数，函数内部发生异常</span>
    <span class="n">this_fails</span><span class="p">()</span>
<span class="k">except</span> <span class="ne">ZeroDivisionError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;调用函数时发生的异常:&#39;</span><span class="p">,</span> <span class="n">err</span><span class="p">)</span></code></pre></td></tr></table>
</div>
</div>
<p>执行以上程序会输出如下结果：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="err">调用函数时发生的异常</span><span class="p">:</span> <span class="n">division</span> <span class="n">by</span> <span class="n">zero</span></code></pre></td></tr></table>
</div>
</div>
<hr />

<p>转载请注明本网址。</p>

    </div>

    
<footer class="post-footer">
      <div class="post-tags">
          <a href="/tags/python%E6%95%99%E7%A8%8B/">python教程</a>
          </div>
      <nav class="post-nav">
        <a class="prev" href="/post/python/python%E6%95%99%E7%A8%8B066-%E5%86%85%E7%BD%AE%E5%BC%82%E5%B8%B8%E7%B1%BB-%E8%87%AA%E5%AE%9A%E4%B9%89%E5%BC%82%E5%B8%B8/">
            <i class="iconfont icon-left"></i>
            <span class="prev-text nav-default">Python教程066 内置异常类 自定义异常</span>
            <span class="prev-text nav-mobile">上一篇</span>
          </a>
        <a class="next" href="/post/python/python%E6%95%99%E7%A8%8B064-%E5%BA%8F%E5%88%97%E5%8C%96%E4%B8%8E%E5%8F%8D%E5%BA%8F%E5%88%97%E5%8C%96/">
            <span class="next-text nav-default">Python教程064 序列化与反序列化</span>
            <span class="next-text nav-mobile">下一篇</span>
            <i class="iconfont icon-right"></i>
          </a>
      </nav>
    </footer>
  </article>
        </div>
        
  
  

  

      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="social-links">
      <a href="https://github.com/inuwashi123" class="iconfont icon-github" title="github"></a>
  <a href="https://inuwashi123.github.io/index.xml" type="application/rss+xml" class="iconfont icon-rss" title="rss"></a>
</div>

<div class="copyright">
  <span class="power-by">
    由 <a class="hexo-link" href="https://gohugo.io">Hugo</a> 强力驱动
  </span>
  <span class="division">|</span>
  <span class="theme-info">
    主题 - 
    <a class="theme-link" href="https://github.com/olOwOlo/hugo-theme-even">Even</a>
  </span>

  

  <span class="copyright-year">
    &copy; 
    2019 - 
    2020
    <span class="heart">
      <i class="iconfont icon-heart"></i>
    </span>
    <span class="author">inuwashi</span>
  </span>
</div>
    </footer>

    <div class="back-to-top" id="back-to-top">
      <i class="iconfont icon-up"></i>
    </div>
  </div>
  
  <script src="https://cdn.jsdelivr.net/npm/jquery@3.2.1/dist/jquery.min.js" integrity="sha256-hwg4gsxgFZhOsEEamdOYGBf13FyQuiTwlAQgxVSNgt4=" crossorigin="anonymous"></script>
  <script src="https://cdn.jsdelivr.net/npm/slideout@1.0.1/dist/slideout.min.js" integrity="sha256-t+zJ/g8/KXIJMjSVQdnibt4dlaDxc9zXr/9oNPeWqdg=" crossorigin="anonymous"></script>
  <script src="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.1.20/dist/jquery.fancybox.min.js" integrity="sha256-XVLffZaxoWfGUEbdzuLi7pwaUJv1cecsQJQqGLe7axY=" crossorigin="anonymous"></script>
<script type="text/javascript" src="/dist/even.26188efa.min.js"></script>








</body>
</html>
